File: frontend/src/shared/api/emailTemplates.ts
import { apiClient } from "./client";

export interface TemplateVariable {
  name: string;
  type: string;
  description: string | null;
  example: string | null;
}

export interface EmailTemplateRead {
  id: string;
  code: string;
  name: string;
  description: string | null;
  subject: string;
  body_html: string;
  body_text: string | null;
  variables: TemplateVariable[];
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface EmailTemplateCreate {
  code: string;
  name: string;
  description?: string | null;
  subject: string;
  body_html: string;
  body_text?: string | null;
  variables?: TemplateVariable[];
}

export interface EmailTemplateUpdate {
  name?: string;
  description?: string | null;
  subject?: string;
  body_html?: string;
  body_text?: string | null;
  variables?: TemplateVariable[];
}

export interface EmailTemplatePreviewResponse {
  subject: string;
  body_html: string;
  body_text: string | null;
}

export async function listEmailTemplates(): Promise<EmailTemplateRead[]> {
  const { data } = await apiClient.get<EmailTemplateRead[]>("/email-templates");
  return data;
}

export async function createEmailTemplate(body: EmailTemplateCreate): Promise<EmailTemplateRead> {
  const { data } = await apiClient.post<EmailTemplateRead>("/email-templates", body);
  return data;
}

export async function updateEmailTemplate(id: string, body: EmailTemplateUpdate): Promise<EmailTemplateRead> {
  const { data } = await apiClient.patch<EmailTemplateRead>(`/email-templates/${id}`, body);
  return data;
}

export async function deleteEmailTemplate(id: string): Promise<void> {
  await apiClient.delete(`/email-templates/${id}`);
}

export async function previewEmailTemplate(
  id: string,
  context: Record<string, unknown>,
): Promise<EmailTemplatePreviewResponse> {
  const { data } = await apiClient.post<EmailTemplatePreviewResponse>(
    `/email-templates/${id}/preview`,
    { context },
  );
  return data;
}

File: backend/app/schemas/email_templates.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class TemplateVariable(BaseModel):
    name: str
    type: str = "string"
    description: str | None = None
    example: str | None = None


class EmailTemplateRead(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    description: str | None
    subject: str
    body_html: str
    body_text: str | None
    variables: list[dict[str, Any]]
    is_system: bool
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class EmailTemplateCreate(BaseModel):
    code: str = Field(min_length=1, max_length=128, pattern=r"^[a-z0-9_]+$")
    name: str = Field(min_length=1, max_length=256)
    description: str | None = None
    subject: str = Field(min_length=1)
    body_html: str = Field(min_length=1)
    body_text: str | None = None
    variables: list[dict[str, Any]] = Field(default_factory=list)


class EmailTemplateUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=256)
    description: str | None = None
    subject: str | None = None
    body_html: str | None = None
    body_text: str | None = None
    variables: list[dict[str, Any]] | None = None


class EmailTemplatePreviewRequest(BaseModel):
    context: dict[str, Any] = Field(default_factory=dict)


class EmailTemplatePreviewResponse(BaseModel):
    subject: str
    body_html: str
    body_text: str | None
